examples/screenshot.c

```diff
@@ -0,0 +1,288 @@
+		void *d = data + (output->offset_y - min_y) * buffer_stride +
+			(output->offset_x - min_x) * 4;
+
+		for (int i = 0; i < output->height; i++) {
+			memcpy(d, s, output_stride);
+			d += buffer_stride;
+			s += output_stride;
+		}
+
+		free(output);
+	}
+
+	argb_to_rgba(data, height, buffer_stride);
+
+	char size[10 + 1 + 10 + 2 + 1]; // int32_t are max 10 digits
+	sprintf(size, "%dx%d+0", width, height);
+
+	int fd[2];
+	pipe(fd);
+
+	pid_t child = fork();
+	if (child < 0) {
+		fprintf(stderr, "fork() failed\n");
+		exit(EXIT_FAILURE);
+	} else if (child != 0) {
+		close(fd[0]);
+		write(fd[1], data, buffer_stride * height);
+		close(fd[1]);
+		free(data);
+		waitpid(child, NULL, 0);
+	} else {
+		close(fd[1]);
+		if (dup2(fd[0], 0) != 0) {
+			fprintf(stderr, "cannot dup the pipe\n");
+			exit(EXIT_FAILURE);
+		}
+		close(fd[0]);
+		execlp("convert", "convert", "-depth", "8", "-size", size, "rgba:-",
+			"-alpha", "opaque", filename, NULL);
+		fprintf(stderr, "cannot execute convert\n");
+		exit(EXIT_FAILURE);
+	}
+}
+
+static int set_buffer_size(int *width, int *height) {
+	struct screenshooter_output *output;
+	min_x = min_y = INT_MAX;
+	max_x = max_y = INT_MIN;
+	int position = 0;
+
+	wl_list_for_each_reverse(output, &output_list, link) {
+		output->offset_x = position;
+		position += output->width;
+	}
+
+	wl_list_for_each(output, &output_list, link) {
+		min_x = MIN(min_x, output->offset_x);
+		min_y = MIN(min_y, output->offset_y);
+		max_x = MAX(max_x, output->offset_x + output->width);
+		max_y = MAX(max_y, output->offset_y + output->height);
+	}
+
+	if (max_x <= min_x || max_y <= min_y) {
+		return -1;
+	}
+
+	*width = max_x - min_x;
+	*height = max_y - min_y;
+
+	return 0;
+}
+
+int main(int argc, char *argv[]) {
+	struct wl_display * display = wl_display_connect(NULL);
+	if (display == NULL) {
+		fprintf(stderr, "failed to create display: %m\n");
+		return -1;
+	}
+
+	wl_list_init(&output_list);
+	struct wl_registry *registry = wl_display_get_registry(display);
+	wl_registry_add_listener(registry, &registry_listener, NULL);
+	wl_display_dispatch(display);
+	wl_display_roundtrip(display);
+
+	if (screenshooter == NULL) {
+		fprintf(stderr, "display doesn't support screenshooter\n");
+		return -1;
+	}
+
+	int width, height;
+	if (set_buffer_size(&width, &height)) {
+		fprintf(stderr, "cannot set buffer size\n");
+		return -1;
+	}
+
+	struct screenshooter_output *output;
+	wl_list_for_each(output, &output_list, link) {
+		if (output->width == 0 || output->height == 0) {
+			continue;
+		}
+
+		output->buffer = create_shm_buffer(output->width, output->height, &output->data);
+		if (output->buffer == NULL) {
+			return -1;
+		}
+		struct orbital_screenshot *screenshot = orbital_screenshooter_shoot(
+			screenshooter, output->output, output->buffer);
+		orbital_screenshot_add_listener(screenshot, &screenshot_listener, screenshot);
+		buffer_copy_done = 0;
+		while (!buffer_copy_done)
+			wl_display_roundtrip(display);
+	}
```

Comment:
Wait, how does this work if there are several outputs?

Reply:
I guess you're trying to stitch them together?

Reply:
This part is from weston's screenshooter. If there are multiple outputs, take a screenshot of each one, and then put them together in one image according to each output's position and size. `set_buffer_size` computes the global buffer size, `write_image` combines each screenshot.

Reply:
Gotcha.